Fix Nested Array TupleData Serialization

Maxxen opened the pull request:
This is a follow-up PR to #10483 that fixes gathering nested arrays with complex types and NULLs in the tuple-format.

Since the tuple format uses a more complicated heap serialization format with flattening and compaction of regular lists, it is extremely difficult to adapt it to array types while also upholding their invariants regarding validity and placement of child elements. This PR instead introduces a workaround where we gather the serialized array as if it was a LIST, and then perform a cast in a new top-level gather function. This isn't super elegant but will at least work for arbitrary nested arrays and a lot of the complexity and scaffolding introduced in the previous attempt can be cleaned up (no more crazy templating). 

We can attempt to optimize this and introduce a proper heap format for arrays in the future, although it would create somewhat of a combinatorial explosion.

Maxxen committed:
gather arrays as list, then cast

lnkuiper commented on src/common/types/row/tuple_data_scatter_gather.cpp:
```diff
@@ -1448,36 +1286,103 @@ static void TupleDataCollectionWithinCollectionGather(const TupleDataLayout &lay
+                                                 const SelectionVector &target_sel, Vector &dummy_vector,
+                                                 const vector<TupleDataGatherFunction> &child_functions) {
+
+	Vector cast_vector(TypeReplaceArrayWithList(target.GetType()));
```
Instead of computing `TypeReplaceArrayWithList` and allocating the `cast_vector` (which is potentially deeply nested) for every scatter/gather, would it be possible to cache the type/vector and re-use them?

I think that `TypeReplaceArrayWithList` then only need to be called when initializing the `TupleDataGatherFunction` (to get the scatter/gather for `LIST` instead), and when initializing `TupleDataChunkState` (the allocated `cast_vector` can live there).

This would already make scatter/gather much more efficient.

Maxxen replied:
Yes, I initially tried to cache it in the tuple layout. The problem is that we don't always call gather with a TupleDataChunkState readily available. (perfect_hash_join_executor, radix_partitioning, join_hashtable)

lnkuiper replied:
1. `perfect_hash_join_executor.cpp` won't be performed if there is an `ARRAY`: `CheckForPerfectJoinOpt` in `plan_comparison_join.cpp
2. `radix_partitioning.cpp` only gathers a column of type `HASH`, so no problem there
3. `join_hashtable.cpp` this one is a problem

For (3.) we could consider making the cached type/vector an `optional_ptr`, so that scanning/appending is more efficient, but it can't be done for random gathers.

I understand this is a bugfix PR, so it's also OK if we leave this optimization for later.

Maxxen replied:
Cool! Yeah I think I can get something like that going for 3. Ill give it a shot!

lnkuiper reviewed (changes requested):
Great idea, let's avoid the combinatorial explosion. I have one request to make it more efficient:

Maxxen committed:
wip try to get caching working

Maxxen committed:
Merge branch 'main' into array-tupledata-fix

Maxxen committed:
wip

Maxxen committed:
disable cache for now just to test CI overnight

Maxxen committed:
rework list to array casts

Maxxen committed:
Fix initialization of cache vectors

Maxxen committed:
Merge branch 'main' into array-tupledata-fix

Maxxen commented:
Additional changes
- Turns out that list->array casts hard an edge case as well, but that's fixed now. There's a new fast path when no lists are null, but the slow path is now slower. (we have to copy the individual lists one by one)
- We now use cached cast vectors to reuse across gathers in some cases (radix partition scan, scan, append)
- Also added a fast-path to avoid recursing down the logical types to check for arrays if the type is not nested.
- Added utility functions to check if a logical type contains another type (or a type matching some predicate) recursively.

Mytherin commented:
Thanks for the PR! I've played around with this for a bit and it seems pretty robust to me. @lnkuiper is this ready to merge?
